### ISSUE TITLE
ROMIO: icc treats "conflicting types" as a warning

### ISSUE THREAD

**Issue by wkliao** [pmodels/mpich#3748]
Function [PAC_TEST_NEEDS_CONST](https://github.com/pmodels/mpich/blob/2407581300b18d6de97ea48e144c0d20f0f13d75/confdb/aclocal_romio.m4#L838-L850
) defined in confdb/aclocal_romio to test whether const declarations is needed for MPI-IO function arguments can fail. For example, gcc treats "conflicting types" in argument declaration as an error, but Intel icc treats it as a warning. Thus, for icc this test always results in "no". Below are compile messages from icc and gcc.

From icc
```
warning #147: declaration is incompatible with "int MPI_File_delete(const char *, MPI_Info={int})" (declared at line 125 of "/opt/cray/pe/mpt/7.7.3/gni/mpich-intel/16.0/include/mpio.h")
  int MPI_File_delete(char *filename, MPI_Info info) { return (0); }
      ^
```
From gcc
```
error: conflicting types for 'MPI_File_delete'
     int MPI_File_delete(char *filename, MPI_Info info) { return (0); }
         ^
/homes/wkliao/MPICH/include/mpio.h:125:5: note: previous declaration is here
int MPI_File_delete(const char *filename, MPI_Info info);
    ^
1 error generated.
```

**Comment by hzhou** [comment]
Shall we try add `-Werror` to that test?
```
PAC_PUSH_FLAG(CFLAGS)
CFLAGS="$CFLAGS -Werror"
...
PAC_POP_FLAG(CFLAGS)
```

**Comment by wkliao** [comment]
Compile option `-Werror` works for gcc and icc, but may not for others, for example, PGI pgcc.

**Comment by wkliao** [comment]
One solution is to check `MPI_VERSION` defined in mpi.h.
The C `const` modifier was first added in MPI standard 3.0.

**Comment by pavanbalaji** [comment]
I don't think we need to do that.  We should simply force const everywhere, and not try to support pre-3.0 standards.

**Comment by wkliao** [comment]
ROMIO is a special case. It is used by other, if not all, MPI implementations.
Resolving this problem makes ROMIO more portable.
For example, I encountered this problem when building ROMIO stand alone using Cray cc compiler.

**Comment by pavanbalaji** [comment]
Every MPI implementation out there have already moved on to MPI-3.0 (including Cray).  If not, they can easily add const support in their MPI implementation.  Maintaining this support is an unnecessary waste of time.

**Comment by wkliao** [comment]
A patch is provided anyway in #3751

**Comment by wkliao** [comment]
Resolved in #4008
